genienlp/tasks/almond_task.py

```diff
@@ -28,135 +21,309 @@
-
-        """Create dataset objects for splits of the ThingTalk dataset.
-        Arguments:
-            path: path to directory where data splits reside
-            train: The prefix of the train data. Default: 'train'.
-            validation: The prefix of the validation data. Default: 'eval'.
-            test: The prefix of the test data. Default: 'test'.
-            Remaining keyword arguments: Passed to the splits method of
-                Dataset.
-        """
-        
-        train_data = None if train is None else cls(os.path.join(path, train + '.tsv'), **kwargs)
-        validation_data = None if validation is None else cls(os.path.join(path, validation + '.tsv'), **kwargs)
-        test_data = None if test is None else cls(os.path.join(path, test + '.tsv'), **kwargs)
-
-        aux_data = None
-        do_curriculum = kwargs.get('curriculum', False)
-        if do_curriculum:
-            kwargs.pop('curriculum')
-            aux_data = cls(os.path.join(path, 'aux' + '.tsv'), **kwargs)
-        
-        return Split(train=None if train is None else train_data,
-                     eval=None if validation is None else validation_data,
-                     test=None if test is None else test_data,
-                     aux=None if do_curriculum is None else aux_data)
-    
 
 
 class BaseAlmondTask(BaseTask):
     """Base class for the Almond semantic parsing task
         i.e. natural language to formal language (ThingTalk) mapping"""
-
+    
     def __init__(self, name, args):
         super().__init__(name, args)
+        self.args = args
+        no_feature_fields = ['answer']
+        if self.is_contextual():
+            no_feature_fields.append('context')
+        else:
+            no_feature_fields.append('question')
+        self.no_feature_fields = no_feature_fields
+        
         self._almond_has_multiple_programs = args.almond_has_multiple_programs
         self._almond_detokenize_sentence = args.almond_detokenize_sentence
+        self._almond_thingtalk_version = args.almond_thingtalk_version
+        
+        self.almond_domains = args.almond_domains
+        self.all_schema_types = set()
+
+        if args.do_ner:
+            self.unk_id = args.features_default_val[0]
+            self.TTtype2DBtype = dict()
+            for domain in self.almond_domains:
+                self.TTtype2DBtype.update(DOMAIN_TYPE_MAPPING[domain])
+            self.DBtype2TTtype = {v: k for k, v in self.TTtype2DBtype.items()}
+            self._init_db()
+
+    def _init_db(self):
+        if self.args.database_type in ['json']:
+            with open(os.path.join(self.args.database_dir, 'es_material/canonical2type.json'), 'r') as fin:
+                canonical2type = ujson.load(fin)
+            with open(os.path.join(self.args.database_dir, 'es_material/type2id.json'), 'r') as fin:
+                type2id = ujson.load(fin)
+            all_canonicals = marisa_trie.Trie(canonical2type.keys())
```

Comment:
The Marisa Trie library is designed to load/write binary files directly from disk - that's how it becomes memory efficient. Please refactor.

Reply:
So canonical2type is about 4G. I can create another json containing only canonical2type keys or I can save `canonical2type.keys()` to memory on the fly and load with marisa_trie. In the first case, we lose disk memory and the second approach might be slower than the current implementation? What do you think?

Reply:
Flagged for next PR.